File: backend/schemas/environment.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

class EnvironmentCreate(BaseModel):
    src: str
    name: str = ""
    os: str = "Windows11"
    user_agent: str = ""
    group: str = "未分组"
    notes: str = ""
    proxy_mode: str = ""
    proxy_type: str = "no"
    proxy_ip_channel: str = ""
    proxy_account_platform: str = ""
    proxy_tabs: str = ""
    webrtc: str = "禁用"
    timezone_mode: str = "自定义"
    timezone: str = "Asia/Shanghai"
    language_mode: str = "真实"
    language: str = "zh-CN"
    resolution_mode: str = "真实"
    resolution: str = ""
    webgl_metadata: str = ""
    webgl_vendor: str = "Google Inc."
    webgl_renderer: str = ""
    canvas: str = "real"
    webgl_image: str = "real"
    audiocontext: str = "real"
    media_devices: str = "real"
    clientrects: str = "real"
    plugin: str = "real"
    cpu_mode: str = "真实"
    cpu: str = "real"
    ram_mode: str = "真实"
    ram: str = "real"
    device_name_mode: str = "真实"
    device_name: str = "real"
    mac_mode: str = "真实"
    mac_address: str = "real"
    do_not_track: str = "默认"
    port_scan_mode: str = "启用"
    port_scan_ports: str = ""
    hardware_acceleration: str = ""
    tls: str = "真实"
    startup_args: str = ""

class EnvironmentUpdate(BaseModel):
    name: Optional[str] = None
    os: Optional[str] = None
    user_agent: Optional[str] = None
    group: Optional[str] = None
    notes: Optional[str] = None
    proxy_mode: Optional[str] = None
    proxy_type: Optional[str] = None
    proxy_ip_channel: Optional[str] = None
    proxy_account_platform: Optional[str] = None
    proxy_tabs: Optional[str] = None
    webrtc: Optional[str] = None
    timezone_mode: Optional[str] = None
    timezone: Optional[str] = None
    language_mode: Optional[str] = None
    language: Optional[str] = None
    resolution_mode: Optional[str] = None
    resolution: Optional[str] = None
    webgl_metadata: Optional[str] = None
    webgl_vendor: Optional[str] = None
    webgl_renderer: Optional[str] = None
    canvas: Optional[str] = None
    webgl_image: Optional[str] = None
    audiocontext: Optional[str] = None
    media_devices: Optional[str] = None
    clientrects: Optional[str] = None
    plugin: Optional[str] = None
    cpu_mode: Optional[str] = None
    cpu: Optional[str] = None
    ram_mode: Optional[str] = None
    ram: Optional[str] = None
    device_name_mode: Optional[str] = None
    device_name: Optional[str] = None
    mac_mode: Optional[str] = None
    mac_address: Optional[str] = None
    do_not_track: Optional[str] = None
    port_scan_mode: Optional[str] = None
    port_scan_ports: Optional[str] = None
    hardware_acceleration: Optional[str] = None
    tls: Optional[str] = None
    startup_args: Optional[str] = None

class EnvironmentListItem(BaseModel):
    id: int
    src: str
    name: str
    os: str
    user_agent: str
    group: str
    notes: str
    proxy_type: str
    proxy_ip_channel: str
    
    class Config:
        from_attributes = True

class GroupInfo(BaseModel):
    name: str
    count: int

File: frontend/src/types/environment.ts
/** 操作系统类型 */
export type OSType = 'Windows11' | 'Windows10'

/** 代理类型 */
export type ProxyType = 'no' | 'http' | 'socks4' | 'socks5'

/** WebRTC 模式 */
export type WebRTCMode = '转发' | '替换' | '真实' | '禁用'

/** 时区设置模式 */
export type SettingMode = '真实' | '自定义'

/** 噪音开关字段 */
export type NoiseField = 'canvas' | 'webgl_image' | 'audiocontext' | 'media_devices' | 'clientrects' | 'plugin'

/** 环境配置 - 完整版 (从原始 JS 迁移) */
export interface EnvironmentConfig {
  /** 用户 ID */
  user_id: string
  /** 环境唯一标识 (SRC-timestamp-random) */
  src: string
  /** 环境名称 */
  name: string
  /** 操作系统 */
  os: OSType
  /** User-Agent */
  user_agent: string
  /** 分组 */
  group: string
  /** 备注 */
  notes: string

  // === 代理配置 ===
  proxy_mode: string
  proxy_type: string
  proxy_ip_channel: string
  proxy_account_platform: string
  proxy_tabs: string

  // === 指纹特征 ===
  webrtc: WebRTCMode | string
  timezone_mode: SettingMode | string
  timezone: string
  language_mode: SettingMode | string
  language: string
  resolution_mode: string
  resolution: string
  webgl_metadata: string
  webgl_vendor: string
  webgl_renderer: string

  // === 噪音特征 ===
  canvas: string       // 'noise' | 'real'
  webgl_image: string
  audiocontext: string
  media_devices: string
  clientrects: string
  plugin: string

  // === 硬件伪装 ===
  cpu_mode: SettingMode | string
  cpu: string
  ram_mode: SettingMode | string
  ram: string
  device_name_mode: SettingMode | string
  device_name: string
  mac_mode: SettingMode | string
  mac_address: string

  // === 隐私特性 ===
  do_not_track: string
  port_scan_mode: string
  port_scan_ports: string
  hardware_acceleration: string
  tls: string
  startup_args: string
}

/** 指纹配置子集 (用于生成 fingerprint JSON) */
export interface FingerprintConfig {
  src: number
  proxy: string
  url: string
  os: string
  canvas: string
  webgl_image: string
  audiocontext: string
  media_devices: string
  clientrects: string
  plugin: string
  cpu: string
  ram: string
  tls: string
  timezone: string
  resolution: string
  user_agent: string
  gpu: string
  webrtc: string
  language: string
  port_scan_ports: string
  startup_args: string
}

/** 环境列表项 (用于列表展示) */
export interface EnvironmentListItem {
  id: number
  src: string
  name: string
  os: string
  user_agent: string
  group: string
  notes: string
  proxy_type: string
  proxy_ip_channel: string
}

/** 分组信息 */
export interface GroupInfo {
  name: string
  count: number
}

/** 文件同步响应 */
export interface SyncInstance {
  files: {
    file_name: string
    url: string
  }[]
}

export interface SyncResponse {
  status: 'success' | 'error'
  message?: string
  instances: SyncInstance[]
}

/** 删除远程文件请求 */
export interface DeleteRemoteFilesRequest {
  user_id: string
  folder_name: string
  files: string[]
}
